File: challenges/2/solution.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def addTwoNumbers(self, l1, l2):
    
    # recursion
    up = 0
    sum = l1.val + l2.val
    if (sum >= 10):
        sum %= 10
        up = 1
        
    # final state
    if (l1.next == None) and (l2.next == None):
        if(up):
            return ListNode(sum, ListNode(up)) # special case    
        return ListNode(sum)
    
    if (l1.next == None):
        l1.next = ListNode()
    
    if (l2.next == None):
        l2.next = ListNode()
    
    # add one to the next node if >= 10
    if(up):
        l1.next.val +=1 
    return ListNode(sum, self.addTwoNumbers(l1.next, l2.next))

File: challenges/2/solution.ts
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function addTwoNumbers(
  l1: ListNode | null,
  l2: ListNode | null
): ListNode | null {
  const dummy = new ListNode(0);
  let current = dummy;
  let carry = 0;

  while (l1 !== null || l2 !== null) {
    const x = l1 !== null ? l1.val : 0;
    const y = l2 !== null ? l2.val : 0;
    const sum = x + y + carry;
    carry = Math.floor(sum / 10);

    current.next = new ListNode(sum % 10);
    current = current.next;

    if (l1 !== null) l1 = l1.next;
    if (l2 !== null) l2 = l2.next;
  }

  if (carry > 0) {
    current.next = new ListNode(carry);
  }

  return dummy.next;
}
